sinks/grpsink/grpc_stream.go

```diff
@@ -0,0 +1,273 @@
+			// to happen here, so we record entering the Idle state as an error.
+			gs.log.WithFields(logrus.Fields{
+				"name":          gs.name,
+				"target":        gs.target,
+				"chanstate":     gs.grpcConn.GetState().String(),
+				logrus.ErrorKey: fmt.Errorf("gRPC sink became idle; should be nearly impossible"),
+			}).Error("gRPC channel transitioned to idle")
+
+			// With the error sent, now fall through to recreating the stream
+			// as if the channel were ready, as stream creation should force
+			// the channel back out of idle. This isn't ideal - it'll end up
+			// recreating the stream again after re-reaching READY - but it's
+			// not a huge problem, as this case is essentially unreachable.
+			fallthrough
+		case connectivity.Ready:
+			gs.streamMut.Lock()
+			// CAS inside the mutex so that we can't ever get a double-send of the
+			// "first time channel degradation" log error.
+			atomic.CompareAndSwapUint32(&gs.bad, 1, 0)
+			stream, err := gs.sinkClient.SendSpans(context.TODO())
+			if err != nil {
+				// This is a weird case and probably shouldn't be reachable, but
+				// if stream setup fails after recovery, then just wait for a
+				// second and try again.
+				gs.log.WithFields(logrus.Fields{
+					"name":          gs.name,
+					"target":        gs.target,
+					"chanstate":     gs.grpcConn.GetState().String(),
+					logrus.ErrorKey: err,
+				}).Error("Failed to set up a new stream after gRPC channel recovered")
+				time.Sleep(1 * time.Second)
+				gs.streamMut.Unlock()
+				continue
+			}
+			gs.stream = stream
+			gs.streamMut.Unlock()
+			gs.log.WithFields(logrus.Fields{
+				"name":      gs.name,
+				"target":    gs.target,
+				"chanstate": gs.grpcConn.GetState().String(),
+			}).Info("gRPC channel and stream re-established")
+		case connectivity.Connecting:
+			gs.log.WithFields(logrus.Fields{
+				"name":      gs.name,
+				"target":    gs.target,
+				"chanstate": state,
+			}).Info("Attempting to re-establish gRPC channel connection")
+			// Nothing to do here except wait.
+		case connectivity.TransientFailure:
+			gs.log.WithFields(logrus.Fields{
+				"name":      gs.name,
+				"target":    gs.target,
+				"chanstate": state,
+			}).Warn("gRPC channel now in transient failure")
+		case connectivity.Shutdown:
+			// The current design has no path to actually shutting down the
+			// connection, so this should be unreachable.
+			return
+		}
+		gs.grpcConn.WaitForStateChange(context.Background(), state)
+	}
+}
+
+// Name returns this sink's name. As the gRPC sink is generic, it's expected
+// that this is set via configuration and injected.
+func (gs *GRPCStreamingSpanSink) Name() string {
+	return gs.name
+}
+
+// Ingest takes in a span and streams it over gRPC to the connected server.
+func (gs *GRPCStreamingSpanSink) Ingest(ssfSpan *ssf.SSFSpan) error {
+	if err := protocol.ValidateTrace(ssfSpan); err != nil {
+		return err
+	}
+	// TODO(sdboyer) validation of span, e.g. time bounds like in datadog sink?
```

Comment:
what do you mean by that TODO? (`ValidateTrace` should take care of identifying only trace spans...)

Reply:
https://github.com/stripe/veneur/blob/a634bb6343326cf2f28a7b86ddebc7291cd5acdd/sinks/datadog/datadog.go#L334-L342